File: c-lib/src/ws.c
/* Server-side WebSocket adapter — see ws.h and PROTOCOL.md §10. */
#include "ws.h"

#include <stdlib.h>
#include <string.h>
#include <strings.h>

#include <openssl/evp.h>
#include <openssl/sha.h>

#define WS_GUID "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

/* Opcodes (RFC 6455 §5.2). */
#define OP_CONT 0x0
#define OP_TEXT 0x1
#define OP_BIN 0x2
#define OP_CLOSE 0x8
#define OP_PING 0x9
#define OP_PONG 0xA

#define MAX_REQUEST_BYTES 8192

void aiomsg_ws_init(aiomsg_ws *w) { memset(w, 0, sizeof(*w)); }

void aiomsg_ws_free(aiomsg_ws *w) {
    free(w->in);
    free(w->out);
    memset(w, 0, sizeof(*w));
}

/* --- growable byte buffers ------------------------------------------------ */

static int buf_append(uint8_t **buf, size_t *len, size_t *cap, const uint8_t *data, size_t n) {
    if (*len + n > *cap) {
        size_t nc = *cap ? *cap : 256;
        while (nc < *len + n) {
            nc *= 2;
        }
        uint8_t *nb = realloc(*buf, nc);
        if (!nb) {
            return -1;
        }
        *buf = nb;
        *cap = nc;
    }
    memcpy(*buf + *len, data, n);
    *len += n;
    return 0;
}

/* --- handshake ------------------------------------------------------------ */

void aiomsg_ws_compute_accept(const char *key, size_t key_len, char *out) {
    unsigned char cat[512];
    size_t n = key_len < sizeof(cat) - sizeof(WS_GUID) ? key_len : 0;
    memcpy(cat, key, n);
    memcpy(cat + n, WS_GUID, sizeof(WS_GUID) - 1);
    unsigned char digest[SHA_DIGEST_LENGTH];
    SHA1(cat, n + sizeof(WS_GUID) - 1, digest);
    EVP_EncodeBlock((unsigned char *)out, digest, SHA_DIGEST_LENGTH); /* writes 28 chars + NUL */
}

/* Case-insensitive lookup of a header value into `val` (NUL-terminated, capped).
 * Returns 1 if found. `req` is the request text (NUL not required); scans lines
 * separated by CRLF. */
static int header_get(const char *req, size_t req_len, const char *name,
                      char *val, size_t val_cap) {
    size_t name_len = strlen(name);
    size_t i = 0;
    /* Skip the request line. */
    while (i < req_len && !(req[i] == '\r' && i + 1 < req_len && req[i + 1] == '\n')) {
        i++;
    }
    i += 2;
    while (i < req_len) {
        size_t start = i;
        while (i < req_len && !(req[i] == '\r' && i + 1 < req_len && req[i + 1] == '\n')) {
            i++;
        }
        size_t line_len = i - start;
        i += 2;
        if (line_len == 0) {
            break; /* blank line: end of headers */
        }
        const char *colon = memchr(req + start, ':', line_len);
        if (!colon) {
            continue;
        }
        size_t hn = (size_t)(colon - (req + start));
        if (hn == name_len && strncasecmp(req + start, name, name_len) == 0) {
            const char *v = colon + 1;
            size_t vlen = line_len - hn - 1;
            while (vlen && (*v == ' ' || *v == '\t')) {
                v++;
                vlen--;
            }
            while (vlen && (v[vlen - 1] == ' ' || v[vlen - 1] == '\t')) {
                vlen--;
            }
            if (vlen >= val_cap) {
                vlen = val_cap - 1;
            }
            memcpy(val, v, vlen);
            val[vlen] = '\0';
            return 1;
        }
    }
    return 0;
}

/* Does the comma-separated token list contain `token` (case-insensitive)? */
static int list_has_token(const char *list, const char *token) {
    size_t tl = strlen(token);
    const char *p = list;
    while (*p) {
        while (*p == ' ' || *p == ',' || *p == '\t') {
            p++;
        }
        const char *start = p;
        while (*p && *p != ',') {
            p++;
        }
        size_t len = (size_t)(p - start);
        while (len && (start[len - 1] == ' ' || start[len - 1] == '\t')) {
            len--;
        }
        if (len == tl && strncasecmp(start, token, tl) == 0) {
            return 1;
        }
    }
    return 0;
}

/* base64-decoded length of `s` (no decode), or -1 if not valid base64 length. */
static int b64_decoded_len(const char *s) {
    size_t n = strlen(s);
    if (n == 0 || n % 4 != 0) {
        return -1;
    }
    int pad = 0;
    if (s[n - 1] == '=') {
        pad++;
    }
    if (s[n - 2] == '=') {
        pad++;
    }
    return (int)(n / 4 * 3) - pad;
}

static uint8_t *make_response(const char *text, size_t *out_len) {
    size_t n = strlen(text);
    uint8_t *r = malloc(n);
    if (r) {
        memcpy(r, text, n);
        *out_len = n;
    }
    return r;
}

int aiomsg_ws_handshake_response(const uint8_t *req, size_t req_len,
                                 uint8_t **resp, size_t *resp_len) {
    const char *r = (const char *)req;
    char val[256];

    /* Request line: GET <path> HTTP/1.1 */
    if (req_len > MAX_REQUEST_BYTES || req_len < 4 || memcmp(r, "GET ", 4) != 0) {
        *resp = make_response("HTTP/1.1 400 Bad Request\r\n\r\n", resp_len);
        return -1;
    }
    if (!header_get(r, req_len, "Upgrade", val, sizeof(val)) || strcasecmp(val, "websocket") != 0) {
        *resp = make_response("HTTP/1.1 400 Bad Request\r\n\r\n", resp_len);
        return -1;
    }
    if (!header_get(r, req_len, "Connection", val, sizeof(val)) || !list_has_token(val, "upgrade")) {
        *resp = make_response("HTTP/1.1 400 Bad Request\r\n\r\n", resp_len);
        return -1;
    }
    if (!header_get(r, req_len, "Sec-WebSocket-Version", val, sizeof(val)) || strcmp(val, "13") != 0) {
        *resp = make_response("HTTP/1.1 426 Upgrade Required\r\nSec-WebSocket-Version: 13\r\n\r\n",
                              resp_len);
        return -1;
    }
    if (!header_get(r, req_len, "Sec-WebSocket-Key", val, sizeof(val)) || b64_decoded_len(val) != 16) {
        *resp = make_response("HTTP/1.1 400 Bad Request\r\n\r\n", resp_len);
        return -1;
    }

    char accept[32];
    aiomsg_ws_compute_accept(val, strlen(val), accept);
    char out[256];
    int n = snprintf(out, sizeof(out),
                     "HTTP/1.1 101 Switching Protocols\r\n"
                     "Upgrade: websocket\r\n"
                     "Connection: Upgrade\r\n"
                     "Sec-WebSocket-Accept: %s\r\n"
                     "\r\n",
                     accept);
    *resp = malloc((size_t)n);
    if (!*resp) {
        *resp_len = 0;
        return -1;
    }
    memcpy(*resp, out, (size_t)n);
    *resp_len = (size_t)n;
    return 0;
}

/* --- frame layer ---------------------------------------------------------- */

uint8_t *aiomsg_ws_encode(int opcode, const uint8_t *payload, size_t len, size_t *out_len) {
    size_t header = 2;
    /* TODO(frame-size): enforce a configurable maximum frame length */
    if (len >= 65536) {
        header = 10;
    } else if (len >= 126) {
        header = 4;
    }
    uint8_t *f = malloc(header + len);
    if (!f) {
        return NULL;
    }
    f[0] = (uint8_t)(0x80 | opcode); /* FIN + opcode, RSV=0 */
    if (header == 2) {
        f[1] = (uint8_t)len;
    } else if (header == 4) {
        f[1] = 126;
        f[2] = (uint8_t)(len >> 8);
        f[3] = (uint8_t)len;
    } else {
        f[1] = 127;
        for (int i = 0; i < 8; i++) {
            f[2 + i] = (uint8_t)(len >> (56 - 8 * i));
        }
    }
    if (len) {
        memcpy(f + header, payload, len);
    }
    *out_len = header + len;
    return f;
}

/* Queue a server frame into w->out. */
static int queue_frame(aiomsg_ws *w, int opcode, const uint8_t *payload, size_t len) {
    size_t flen;
    uint8_t *f = aiomsg_ws_encode(opcode, payload, len, &flen);
    if (!f) {
        return -1;
    }
    int rc = buf_append(&w->out, &w->out_len, &w->out_cap, f, flen);
    free(f);
    return rc;
}

/* Queue a close frame with `code` (once) and signal connection close. */
static int ws_fail(aiomsg_ws *w, uint16_t code) {
    if (!w->close_sent) {
        w->close_sent = 1;
        uint8_t body[2] = {(uint8_t)(code >> 8), (uint8_t)code};
        queue_frame(w, OP_CLOSE, body, 2);
    }
    return -1;
}

int aiomsg_ws_feed(aiomsg_ws *w, const uint8_t *data, size_t len, aiomsg_decoder *dec) {
    if (buf_append(&w->in, &w->in_len, &w->in_cap, data, len) != 0) {
        return -1;
    }
    size_t off = 0;
    for (;;) {
        if (w->in_len - off < 2) {
            break;
        }
        uint8_t *p = w->in + off;
        uint8_t b0 = p[0], b1 = p[1];
        if (b0 & 0x70) {
            return ws_fail(w, 1002); /* RSV bits set */
        }
        int opcode = b0 & 0x0f;
        int fin = b0 & 0x80;
        if (!(b1 & 0x80)) {
            return ws_fail(w, 1002); /* client frames MUST be masked */
        }
        uint64_t plen = b1 & 0x7f;
        size_t hdr = 2;
        if (plen == 126) {
            if (w->in_len - off < 4) {
                break;
            }
            plen = ((uint64_t)p[2] << 8) | p[3];
            hdr = 4;
        } else if (plen == 127) {
            if (w->in_len - off < 10) {
                break;
            }
            if (p[2] & 0x80) {
                return ws_fail(w, 1002); /* 64-bit length MSB must be 0 */
            }
            /* TODO(frame-size): enforce a configurable maximum frame length */
            plen = 0;
            for (int i = 0; i < 8; i++) {
                plen = (plen << 8) | p[2 + i];
            }
            hdr = 10;
        }
        if (opcode >= 0x8 && (!fin || plen > 125)) {
            return ws_fail(w, 1002); /* control frames: FIN=1, <=125 bytes */
        }
        if (w->in_len - off < hdr + 4 + plen) {
            break; /* need mask (4) + payload */
        }
        uint8_t *mask = p + hdr;
        uint8_t *masked = p + hdr + 4;
        for (uint64_t i = 0; i < plen; i++) {
            masked[i] ^= mask[i & 3];
        }
        off += hdr + 4 + (size_t)plen;

        if (opcode == OP_BIN || opcode == OP_CONT) {
            if (aiomsg_decoder_push(dec, masked, (size_t)plen) != 0) {
                return -1;
            }
        } else if (opcode == OP_PING) {
            if (queue_frame(w, OP_PONG, masked, (size_t)plen) != 0) {
                return -1;
            }
        } else if (opcode == OP_PONG) {
            /* ignore */
        } else if (opcode == OP_TEXT) {
            return ws_fail(w, 1003); /* text is not valid for aiomsg */
        } else if (opcode == OP_CLOSE) {
            uint16_t code = plen >= 2 ? (uint16_t)((masked[0] << 8) | masked[1]) : 1000;
            if (!w->close_sent) {
                w->close_sent = 1;
                uint8_t body[2] = {(uint8_t)(code >> 8), (uint8_t)code};
                queue_frame(w, OP_CLOSE, body, 2);
            }
            /* Drop consumed bytes for tidiness, then signal close. */
            memmove(w->in, w->in + off, w->in_len - off);
            w->in_len -= off;
            return -1;
        } else {
            return ws_fail(w, 1002); /* unknown opcode */
        }
    }
    if (off) {
        memmove(w->in, w->in + off, w->in_len - off);
        w->in_len -= off;
    }
    return 0;
}

File: zig-lib/src/ws.zig
//! Server-side WebSocket (RFC 6455) byte-stream adapter for aiomsg **bind**
//! sockets (see `../../PROTOCOL.md` §10).
//!
//! This is a pure, transport-agnostic layer: it turns the payloads of inbound
//! binary WebSocket messages into the concatenated aiomsg byte stream of §2
//! (pushed to the existing `protocol.Decoder`), and frames each outbound aiomsg
//! write as one unmasked binary WebSocket message. WebSocket message boundaries
//! carry no meaning. It never touches a socket itself — the caller (`aiomsg.zig`)
//! feeds it raw bytes and flushes the control-frame bytes it queues — so only the
//! server half of RFC 6455 lives here, with no subprotocol or extension (notably
//! no `permessage-deflate`) ever negotiated.

const std = @import("std");
const Allocator = std.mem.Allocator;
const proto = @import("protocol.zig");

/// The magic GUID from RFC 6455 §1.3, appended to the client key before SHA-1.
const ws_guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

// Opcodes (RFC 6455 §5.2).
const op_cont: u8 = 0x0;
const op_text: u8 = 0x1;
const op_bin: u8 = 0x2;
const op_close: u8 = 0x8;
const op_ping: u8 = 0x9;
const op_pong: u8 = 0xA;

pub const max_request_bytes: usize = 8192;

/// The outcome of feeding inbound bytes: keep serving, or tear the connection
/// down (peer close or protocol error — the close frame, if any, is already
/// queued in `out`).
pub const FeedResult = enum { open, close };

/// Per-connection WebSocket parser state. `in` accumulates partial inbound
/// frames; `out` holds control-frame bytes (pong / close) the caller must write
/// to the peer after each feed.
pub const Ws = struct {
    in: std.ArrayList(u8) = .empty,
    out: std.ArrayList(u8) = .empty,
    close_sent: bool = false,

    pub fn deinit(self: *Ws, gpa: Allocator) void {
        self.in.deinit(gpa);
        self.out.deinit(gpa);
    }
};

// --- handshake -------------------------------------------------------------

/// Write `Sec-WebSocket-Accept` for `key` — `base64(SHA1(key ++ GUID))`,
/// exactly 28 characters (RFC 6455 §4.2.2) — into `out`.
pub fn computeAccept(key: []const u8, out: *[28]u8) void {
    var digest: [std.crypto.hash.Sha1.digest_length]u8 = undefined;
    var h = std.crypto.hash.Sha1.init(.{});
    h.update(key);
    h.update(ws_guid);
    h.final(&digest);
    _ = std.base64.standard.Encoder.encode(out, &digest);
}

/// Case-insensitive lookup of a header value, trimmed of surrounding whitespace.
/// Scans CRLF-separated lines after the request line; returns null if absent.
fn headerGet(req: []const u8, name: []const u8) ?[]const u8 {
    var it = std.mem.splitSequence(u8, req, "\r\n");
    _ = it.next(); // skip the request line
    while (it.next()) |line| {
        if (line.len == 0) break; // blank line: end of headers
        const colon = std.mem.indexOfScalar(u8, line, ':') orelse continue;
        if (std.ascii.eqlIgnoreCase(std.mem.trim(u8, line[0..colon], " \t"), name)) {
            return std.mem.trim(u8, line[colon + 1 ..], " \t");
        }
    }
    return null;
}

/// Does the comma-separated token list contain `token` (case-insensitive)?
fn listHasToken(list: []const u8, token: []const u8) bool {
    var it = std.mem.tokenizeAny(u8, list, ", \t");
    while (it.next()) |tok| {
        if (std.ascii.eqlIgnoreCase(tok, token)) return true;
    }
    return false;
}

/// Number of bytes `key` base64-decodes to, or null if it is not valid base64.
fn base64DecodedLen(key: []const u8) ?usize {
    return std.base64.standard.Decoder.calcSizeForSlice(key) catch null;
}

/// Validate an HTTP upgrade request (bytes through the terminating blank line)
/// and build the response. On success writes a 101 response to `resp` and
/// returns `true`; on any violation writes a 400/426 error response and returns
/// `false`. The caller writes `resp` to the peer in both cases and frees it.
pub fn handshakeResponse(gpa: Allocator, req: []const u8, resp: *[]u8) Allocator.Error!bool {
    const bad_request = "HTTP/1.1 400 Bad Request\r\n\r\n";
    const bad_version = "HTTP/1.1 426 Upgrade Required\r\nSec-WebSocket-Version: 13\r\n\r\n";

    if (req.len > max_request_bytes or req.len < 4 or !std.mem.eql(u8, req[0..4], "GET ")) {
        resp.* = try gpa.dupe(u8, bad_request);
        return false;
    }
    const upgrade = headerGet(req, "Upgrade");
    if (upgrade == null or !std.ascii.eqlIgnoreCase(upgrade.?, "websocket")) {
        resp.* = try gpa.dupe(u8, bad_request);
        return false;
    }
    const connection = headerGet(req, "Connection");
    if (connection == null or !listHasToken(connection.?, "upgrade")) {
        resp.* = try gpa.dupe(u8, bad_request);
        return false;
    }
    const version = headerGet(req, "Sec-WebSocket-Version");
    if (version == null or !std.mem.eql(u8, version.?, "13")) {
        resp.* = try gpa.dupe(u8, bad_version);
        return false;
    }
    const key = headerGet(req, "Sec-WebSocket-Key");
    if (key == null or base64DecodedLen(key.?) != proto.identity_size) {
        resp.* = try gpa.dupe(u8, bad_request);
        return false;
    }

    var accept: [28]u8 = undefined;
    computeAccept(key.?, &accept);
    resp.* = try std.fmt.allocPrint(gpa, "HTTP/1.1 101 Switching Protocols\r\n" ++
        "Upgrade: websocket\r\n" ++
        "Connection: Upgrade\r\n" ++
        "Sec-WebSocket-Accept: {s}\r\n" ++
        "\r\n", .{accept});
    return true;
}

// --- frame layer -----------------------------------------------------------

/// Encode one server->client frame (FIN=1, unmasked, given opcode) into a
/// freshly allocated buffer. opcode 0x2 = binary.
pub fn encode(gpa: Allocator, opcode: u8, payload: []const u8) Allocator.Error![]u8 {
    const len = payload.len;
    // TODO(frame-size): enforce a configurable maximum frame length
    const header: usize = if (len >= 65536) 10 else if (len >= 126) 4 else 2;
    var f = try gpa.alloc(u8, header + len);
    f[0] = 0x80 | opcode; // FIN + opcode, RSV=0
    if (header == 2) {
        f[1] = @intCast(len);
    } else if (header == 4) {
        f[1] = 126;
        f[2] = @intCast((len >> 8) & 0xff);
        f[3] = @intCast(len & 0xff);
    } else {
        f[1] = 127;
        var i: u6 = 0;
        while (i < 8) : (i += 1) {
            f[2 + i] = @intCast((len >> (56 - 8 * @as(u6, i))) & 0xff);
        }
    }
    if (len != 0) @memcpy(f[header..], payload);
    return f;
}

/// Queue a server frame (payload <= 125, so a fixed 2-byte header) into `out`.
fn queueControl(self: *Ws, gpa: Allocator, opcode: u8, payload: []const u8) Allocator.Error!void {
    try self.out.append(gpa, 0x80 | opcode);
    try self.out.append(gpa, @intCast(payload.len));
    try self.out.appendSlice(gpa, payload);
}

/// Queue a close frame with `code` (once) and signal the connection must close.
fn fail(self: *Ws, gpa: Allocator, code: u16) Allocator.Error!FeedResult {
    if (!self.close_sent) {
        self.close_sent = true;
        try queueControl(self, gpa, op_close, &.{ @intCast(code >> 8), @intCast(code & 0xff) });
    }
    return .close;
}

/// Feed raw inbound bytes. Appends decoded binary payload to `dec`, and queues
/// pong / close frames into `out` for the caller to flush. Returns `.open` to
/// keep the connection open, or `.close` when it must be torn down (peer close,
/// or a protocol error — a close frame is queued in `out` first).
pub fn feed(self: *Ws, gpa: Allocator, data: []const u8, dec: *proto.Decoder) Allocator.Error!FeedResult {
    try self.in.appendSlice(gpa, data);
    const buf = self.in.items;
    var off: usize = 0;
    const result: FeedResult = outer: while (true) {
        if (buf.len - off < 2) break :outer .open;
        const p = buf[off..];
        const b0 = p[0];
        const b1 = p[1];
        if (b0 & 0x70 != 0) break :outer try fail(self, gpa, 1002); // RSV bits set
        const opcode = b0 & 0x0f;
        const fin = b0 & 0x80 != 0;
        if (b1 & 0x80 == 0) break :outer try fail(self, gpa, 1002); // client frames MUST be masked
        var plen: u64 = b1 & 0x7f;
        var hdr: usize = 2;
        if (plen == 126) {
            if (buf.len - off < 4) break :outer .open;
            plen = (@as(u64, p[2]) << 8) | p[3];
            hdr = 4;
        } else if (plen == 127) {
            if (buf.len - off < 10) break :outer .open;
            if (p[2] & 0x80 != 0) break :outer try fail(self, gpa, 1002); // 64-bit length MSB must be 0
            // TODO(frame-size): enforce a configurable maximum frame length
            plen = 0;
            var i: usize = 0;
            while (i < 8) : (i += 1) plen = (plen << 8) | p[2 + i];
            hdr = 10;
        }
        if (opcode >= 0x8 and (!fin or plen > 125)) {
            break :outer try fail(self, gpa, 1002); // control frames: FIN=1, <=125 bytes
        }
        if (buf.len - off < hdr + 4 + plen) break :outer .open; // need mask (4) + payload

        const mask = p[hdr .. hdr + 4];
        const masked = p[hdr + 4 .. hdr + 4 + @as(usize, @intCast(plen))];
        for (masked, 0..) |*b, i| b.* ^= mask[i & 3];
        off += hdr + 4 + @as(usize, @intCast(plen));

        switch (opcode) {
            op_bin, op_cont => try dec.push(gpa, masked),
            op_ping => try queueControl(self, gpa, op_pong, masked),
            op_pong => {}, // ignore
            op_text => break :outer try fail(self, gpa, 1003), // text is not valid for aiomsg
            op_close => {
                const code: u16 = if (plen >= 2)
                    (@as(u16, masked[0]) << 8) | masked[1]
                else
                    1000;
                if (!self.close_sent) {
                    self.close_sent = true;
                    try queueControl(self, gpa, op_close, &.{ @intCast(code >> 8), @intCast(code & 0xff) });
                }
                break :outer .close;
            },
            else => break :outer try fail(self, gpa, 1002), // unknown opcode
        }
    };

    if (off != 0) {
        const rem = self.in.items.len - off;
        std.mem.copyForwards(u8, self.in.items[0..rem], self.in.items[off..]);
        self.in.shrinkRetainingCapacity(rem);
    }
    return result;
}

// --- tests -----------------------------------------------------------------

/// Build a masked client frame (as a browser would send). Test-only.
fn maskedFrame(gpa: Allocator, opcode: u8, payload: []const u8) Allocator.Error![]u8 {
    std.debug.assert(payload.len < 126); // tests use short payloads only
    var f = try gpa.alloc(u8, 2 + 4 + payload.len);
    f[0] = 0x80 | opcode;
    f[1] = 0x80 | @as(u8, @intCast(payload.len));
    const mask = [4]u8{ 0x12, 0x34, 0x56, 0x78 };
    @memcpy(f[2..6], &mask);
    for (payload, 0..) |b, i| f[6 + i] = b ^ mask[i & 3];
    return f;
}

test "RFC 6455 accept-key vector" {
    var out: [28]u8 = undefined;
    computeAccept("dGhlIHNhbXBsZSBub25jZQ==", &out);
    try std.testing.expectEqualStrings("s3pPLMBiTxaQ9kYGzzhZRbK+xOo=", &out);
}

test "handshake accepts a valid upgrade" {
    const gpa = std.testing.allocator;
    const req = "GET /chat HTTP/1.1\r\nHost: x\r\nUpgrade: websocket\r\n" ++
        "Connection: keep-alive, Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n" ++
        "Sec-WebSocket-Version: 13\r\n\r\n";
    var resp: []u8 = undefined;
    const ok = try handshakeResponse(gpa, req, &resp);
    defer gpa.free(resp);
    try std.testing.expect(ok);
    try std.testing.expect(std.mem.indexOf(u8, resp, "101 Switching Protocols") != null);
    try std.testing.expect(std.mem.indexOf(u8, resp, "s3pPLMBiTxaQ9kYGzzhZRbK+xOo=") != null);
}

test "handshake rejects a wrong version with 426" {
    const gpa = std.testing.allocator;
    const req = "GET / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\n" ++
        "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 8\r\n\r\n";
    var resp: []u8 = undefined;
    const ok = try handshakeResponse(gpa, req, &resp);
    defer gpa.free(resp);
    try std.testing.expect(!ok);
    try std.testing.expect(std.mem.indexOf(u8, resp, "426") != null);
}

test "feed unmasks a binary message into the decoder" {
    const gpa = std.testing.allocator;
    var w = Ws{};
    defer w.deinit(gpa);
    var dec: proto.Decoder = .{};
    defer dec.deinit(gpa);

    const beat = try proto.frameHeartbeat(gpa);
    defer gpa.free(beat);
    const frame = try maskedFrame(gpa, op_bin, beat);
    defer gpa.free(frame);

    try std.testing.expectEqual(FeedResult.open, try feed(&w, gpa, frame, &dec));
    const e = proto.parseEnvelope(dec.pop().?).?;
    try std.testing.expect(e.type == .heartbeat);
}

test "feed answers a ping with a pong" {
    const gpa = std.testing.allocator;
    var w = Ws{};
    defer w.deinit(gpa);
    var dec: proto.Decoder = .{};
    defer dec.deinit(gpa);

    const frame = try maskedFrame(gpa, op_ping, "hi");
    defer gpa.free(frame);
    try std.testing.expectEqual(FeedResult.open, try feed(&w, gpa, frame, &dec));
    try std.testing.expectEqual(@as(u8, 0x80 | op_pong), w.out.items[0]);
    try std.testing.expectEqualStrings("hi", w.out.items[2..4]);
}

test "feed rejects an unmasked frame with close 1002" {
    const gpa = std.testing.allocator;
    var w = Ws{};
    defer w.deinit(gpa);
    var dec: proto.Decoder = .{};
    defer dec.deinit(gpa);

    var frame = [_]u8{ 0x82, 0x01, 0x00 }; // binary, mask bit clear
    try std.testing.expectEqual(FeedResult.close, try feed(&w, gpa, &frame, &dec));
    try std.testing.expectEqual(@as(u8, 0x88), w.out.items[0]); // close frame
    try std.testing.expectEqual(@as(u8, 0x03), w.out.items[2]); // 1002 = 0x03EA
    try std.testing.expectEqual(@as(u8, 0xEA), w.out.items[3]);
}

test "feed rejects a text frame with close 1003" {
    const gpa = std.testing.allocator;
    var w = Ws{};
    defer w.deinit(gpa);
    var dec: proto.Decoder = .{};
    defer dec.deinit(gpa);

    const frame = try maskedFrame(gpa, op_text, "x");
    defer gpa.free(frame);
    try std.testing.expectEqual(FeedResult.close, try feed(&w, gpa, frame, &dec));
    try std.testing.expectEqual(@as(u8, 0x88), w.out.items[0]);
    try std.testing.expectEqual(@as(u8, 0xEB), w.out.items[3]); // 1003 = 0x03EB
}
